_posts/2016-05-11-android-testing-with-kotlin.md

```diff
@@ -0,0 +1,422 @@
+meaningful and "Add item" button is pressed, the newly added item is populated to the list.
+Furthermore, it is of utmost importance to ensure that all the items that are added to the list
+do not contain any superfluous whitespace characters. For that we need tests of course.
+
+## Unit tests
+
+Let us start off with unit tests for our `Formatter` class shown above. As already mentioned,
+we'll store our unit tests under `src/test/kotlin`. Right there we have a test class
+`FormatterTestKotlin` defined in
+[`FormatterTestKotlin.kt`](https://github.com/GreenhouseCI/kotlin-tests/blob/master/app/src/test/kotlin/com/greenhouseci/kotlin_tests/kotlintests/FormatterTestKotlin.kt):
+
+```ruby
+package com.greenhouseci.kotlin_tests.kotlintests
+
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+
+class FormatterTestKotlin {
+
+    private val formatter: Formatter = Formatter()
+
+    infix fun Any.equals(expected: Any) {
+        assertEquals(expected, this)
+    }
+
+    ...
+
+    @Test
+    @Throws(Exception::class)
+    fun testStrip() {
+        val reference = "my string"
+        val testStrings = arrayOf(
+            reference,
+            " $reference ",
+            " $reference  ",
+            " \t $reference \t ",
+            " \n  $reference \n ",
+            "\tmy \n string\t"
+        )
+        testStrings.map { formatter.strip(it) equals reference }
+    }
+```
+
+A glimpse at the snippet above shows that while it is Kotlin code, there are some
+familiar elements from regular Java already, like package name and import statements.
+Further decomposing it reveals lack of semicolons in the end of the lines and differences
+in variable and function declarations. But also some obvious similarities like code block
+separation by usage of curly brackets and support for annotations.
+
+__Variable and function declaration__
+
+In our sample code we can see that declaring a value can be as simple as
+
+```groovy
+val reference = "my string"
+```
+
+where the type of the object that is assigned is implicitly guessed by compiler.
+Or we can explicitly declare the type of the object by writing
+
+```groovy
+val formatter: Formatter = Formatter()
+```
+
+where the type information follows variable name with colon and precedes the value
+to be assigned.
+
+Kotlin also provides an easy way to distinguish read-only and mutable fields.
+It does so by exposing two types of keywords: `val` for declaring read-only fields
+and `var` for variables that ought to be mutable. In the realm of Java, `val` corresponds
+to the `final` modifier.
+
+> *It is also worth noting how we can assign Java object to a Kotlin variable without
+any further ado.*
+
+Functions are defined using `fun` keyword followed by the function name and parameters
+in parentheses. Return type and argument types are specified using the colon notation
+just as it is with variable declaration. But as Kotlin has first class support for
```

Comment:
It would be more appropriate to use "just as **they are**" since this talks about return type **and** argument types. However, since this does not directly translate to variable declaration, this direct comparison can be avoided with "just as with variable declaration"